### DISCUSSION TITLE
Pynecone Docker configs?

### DISCUSSION THREAD

**Discussion by janzheng** [reflex-dev/reflex#798] [Q&A] [answered]
Hi there,

I'm having trouble building and running Docker locally. It's just a vanilla install, since I'm trying to understand how everything works under the hood.

Partly is I'm not sure what to call things / rename things in the config (e.g. "app"). My app is called "thirdtry" and my current venv and app is all under that name, e.g. `thirdtry/thirdtry/thirdtry.py`.

Which parts should I rename to `thirdtry` and which parts should I keep as `app`? Variations of errors I get include: 
- `pc init` "pc not found" when I build
- when I successfully run Docker locally, I usually get the error:
```
$ next build && next export -o _static
/bin/bash: line 1: next: command not found
error: script "export" exited with code 127
$ next start
/bin/bash: line 1: next: command not found
error: script "prod" exited with code 127
```

Here are my steps:
1. I'm successfully running with `pc init` and `pc run`
2. I'm setting `pconfig.py` like this for Docker:
```
import pynecone as pc

config = pc.Config(
    app_name="thirdtry",
    api_url="0.0.0.0:8000",
    bun_path="/app/.bun/bin/bun",
    db_url="sqlite:///pynecone.db",
)
```
3. Here's my `Dockerfile`:
```
FROM python:3.11-slim as base

RUN adduser --disabled-password pynecone


FROM base as build

WORKDIR /app
ENV VIRTUAL_ENV=/app/venv
RUN python3 -m venv $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

COPY . .

RUN pip install wheel \
    && pip install -r requirements.txt


FROM base as runtime

RUN apt-get update && apt-get install -y \
    curl \
    && curl -fsSL https://deb.nodesource.com/setup_19.x | bash - \
    && apt-get update && apt-get install -y \
    nodejs \
    unzip \
    && rm -rf /var/lib/apt/lists/*

ENV PATH="/app/venv/bin:$PATH"


FROM runtime as init

WORKDIR /app
ENV BUN_INSTALL="/app/.bun"
COPY --from=build /app/ /app/
RUN pc init


FROM runtime

COPY --chown=pynecone --from=init /app/ /app/
USER pynecone
WORKDIR /app

CMD ["pc","run" , "--env", "prod"]

EXPOSE 3000
EXPOSE 8000
```

4. Here's my `requirements.txt`:
```
pynecone
```

Ok I know it's something silly like a misnamed folder or app, but I have no idea how to fix it...

**Comment by janzheng** [comment, marked as answer]
Ok I came up with a Dockerfile that worked!

The problem is that I've never installed some of the Node requirements before. This updated Dockerfile installs npm, yarn, then installs the node requirements.

It's heavily inspired by @tspycher from Discord (repo: https://github.com/tspycher/tspycher)

```
FROM python:3.11-slim as base

RUN adduser --disabled-password pynecone


FROM base as build

WORKDIR /thirdtry
ENV VIRTUAL_ENV=/thirdtry/venv
RUN python3 -m venv $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

COPY . .

RUN pip install wheel \
    && pip install -r requirements.txt


FROM base as runtime

RUN apt-get update && apt-get install -y \
    curl \
    && curl -fsSL https://deb.nodesource.com/setup_19.x | bash - \
    && apt-get update && apt-get install -y \
    nodejs \
    unzip \
    && rm -rf /var/lib/apt/lists/*

RUN npm install -g n
RUN n latest
RUN npm install -g npm@latest
RUN npm install -g yarn

ENV PATH="/thirdtry/venv/bin:$PATH"


FROM runtime as init

WORKDIR /thirdtry
ENV BUN_INSTALL="/thirdtry/.bun"
COPY --from=build /thirdtry/ /thirdtry/
RUN pc init

RUN yarn install --update-checksums --cwd /thirdtry/.web
RUN yarn add next @emotion/react @chakra-ui/react focus-visible


FROM runtime

COPY --chown=pynecone --from=init /thirdtry/ /thirdtry/
USER pynecone
WORKDIR /thirdtry

CMD ["pc","run" , "--env", "prod"]

EXPOSE 3000
EXPOSE 8000
```

**Reply by janzheng** [reply]
Ha nevermind, this causes bizarre "TypeError: Failed to construct 'URL': Invalid URL" errors. Will probably just wait for the `pc deploy` system to get up and running; seems like only static versions work for now.

**Reply by janzheng** [reply]
Ok so apparently it seems like the Pinecone / FastAPI server isn't actually starting on port 8000, that's why I'm getting these errors. No idea how to get it started though.

**Reply by janzheng** [reply]
Confusingly, my deployment has NO problems if I remove the `api_url` line from `pcconfig.py`. I have no idea why it works. In Docker, the output is `ready - started server on 0.0.0.0:3000, url: http://localhost:3000` — which indicates it SHOULDN'T work, right? It says this even when I set the `api_url` which I guess is why the server fails to run, since it ignores the `api_url` and expects it to be on 3000. Something fishy is going on here.

**Reply by picklelo** [reply]
@janzheng What do you set the `api_url` to? When you deploy, make sure you format it like `http://192.182.169.45:8000` for example.

**Comment by picklelo** [comment]
Created this basic guide for self hosting that may be helpful: https://github.com/pynecone-io/pynecone/discussions/802

**Comment by janzheng** [comment]
Ok! I was able to get Pynecone deployed and working on Fly.io as a Docker deployment, and the server responses work!

1. My project is in the the folder `thirdtry`, and the venv is the same name. Named everything "thirdtry" so I didn't run into namespace problems. ("thirdtry" because the previous tried I'd messed up the venv ha)
2. The Dockerfile could probably be simplified, but I had it install yarn, then install some node modules. I explicitly had it install @emotion/react and some other ones because... for some reason the standard install didn't do it. I added `RUN pc export --no-zip` before EXPOSE ing ports — not sure if I can remove these, but it works this way lol
3. The docs were confusing about what was optionally named "app" and what had to be named after my folder / app, so I just changed everywhere "app" existed with "thirdtry" and it seemed to work, including paths
4. On `fly.toml` make sure you add TWO SERVICES, one for `8000` and one for `3000`, but only export `3000` to port 80 and 443. Also note the `[env]` area where it says `FIRST_APP_PORT` and `SECOND_APP_PORT`

I think the configs can be cleaned up quite a bit, but this is working for me at the moment, so I don't want to mess it up by poking the bear too much ;)

`Dockerfile`:
~~~
FROM python:3.11-slim as base

RUN adduser --disabled-password pynecone


FROM base as build

WORKDIR /thirdtry
ENV VIRTUAL_ENV=/thirdtry/venv
ENV BUN_INSTALL="/app/.bun"
RUN python3 -m venv $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

COPY ./assets /thirdtry/assets
COPY . .

RUN pip install wheel \
    && pip install -r requirements.txt


FROM base as runtime

RUN apt-get update && apt-get install -y \
    curl \
    && curl -fsSL https://deb.nodesource.com/setup_19.x | bash - \
    && apt-get update && apt-get install -y \
    nodejs \
    unzip \
    && rm -rf /var/lib/apt/lists/*

RUN npm install -g n
RUN n latest
RUN npm install -g npm@latest
RUN npm install -g yarn

ENV PATH="/thirdtry/venv/bin:$PATH"


FROM runtime as init

WORKDIR /thirdtry
ENV BUN_INSTALL="/thirdtry/.bun"
COPY --from=build /thirdtry/ /thirdtry/
RUN pc init

RUN yarn install --update-checksums --cwd /thirdtry/.web
RUN yarn add next @emotion/react @chakra-ui/react focus-visible


FROM runtime

COPY --chown=pynecone --from=init /thirdtry/ /thirdtry/
USER pynecone
WORKDIR /thirdtry

RUN pc export --no-zip

EXPOSE 8000
EXPOSE 3000
CMD ["pc","run" , "--env", "prod"]
~~~

`pcconfig.py`:
~~~
import pynecone as pc

# Local
# config = pc.Config(
#     app_name="thirdtry",
#     db_url="sqlite:///pynecone.db",
#     env=pc.Env.DEV,
# )

# Fly / Docker
config = pc.Config(
    app_name="thirdtry",
    api_url="http://127.0.0.1:8000",
    bun_path="/thirdtry/.bun/bin/bun",
    db_url="sqlite:///pynecone.db",
)
~~~

`fly.toml`
~~~
# fly.toml file generated for thirdtry on 2023-04-10T21:14:05+10:00

app = "thirdtry"
kill_signal = "SIGINT"
kill_timeout = 5
primary_region = "syd"
processes = []

[env]
  FIRST_APP_PORT = "8000"
  SECOND_APP_PORT = "3000"

[experimental]
  auto_rollback = true


[[services]]
  http_checks = []
  internal_port = 8000
  processes = ["app"]
  protocol = "tcp"
  script_checks = []

  [services.concurrency]
    hard_limit = 25
    soft_limit = 20
    type = "connections"

#  [[services.ports]]
#    force_https = true
#    handlers = ["http"]
#    port = 8000

  [[services.tcp_checks]]
    grace_period = "1s"
    interval = "15s"
    restart_limit = 0
    timeout = "2s"
    
[[services]]
  http_checks = []
  internal_port = 3000
  processes = ["app"]
  protocol = "tcp"
  script_checks = []
  [services.concurrency]
    hard_limit = 25
    soft_limit = 20
    type = "connections"

  [[services.ports]]
    force_https = true
    handlers = ["http"]
    port = 80

  [[services.ports]]
    handlers = ["tls", "http"]
    port = 443

  [[services.tcp_checks]]
    grace_period = "1s"
    interval = "15s"
    restart_limit = 0
    timeout = "2s"



~~~

**Reply by oresttokovenko** [reply]
Forgive me in advance as I'm not that familiar with docker, but is it necessary to write your own dockerfile if Pynecone provides their own image? `pynecone-project:latest`

**Comment by kai-kaushik** [comment]
@janzheng could you please explain how these ports work on the Pynecode (now reflex) app? Deploying the app as been a real nightmare for me. I followed all your instructions and still fail to get a running version of my app. Did you try deploying it in any other way as well such as AWS Elastic Beanstalk?

**Comment by david-thrower** [comment]
I wanted to look for some guidance on 2 issues around docker - k8s based deployment:

1. Building a truly stateless image.
2. Issues surrounding traffic on both port 8000 and port 3000 

Stateless image:

- `In https://github.com/reflex-dev/reflex/blob/14bb662eb1113f4a8dae3b04c6792f9d30243d68/docker-example/README.md`, I am advised to build the container as: `docker build -t reflex-app:latest . --build-arg API_URL=http://app.example.com:8000`.
- The issue I'm seeing is that **--build-arg API_URL=[foo]** makes the container bound to a given host.
- I think this is because Reflex wants to create a SSL certificate for the host.
- In k8s, the SSL is usually preempted by the service mesh within the cluster as mTLS, and the TLS for external traffic is preempted by the Ingress or IngressRoute.
-  Basically, I want to make it run in a pod without TLS and let the cluster manage that.
- Additionally, the stateful build is problematic, because we would need to change the container image entry on the deployment  for each customer we deploy the same app for, as the container is already bound to a given host and build everyone a separate container image. 

Issue with 2 ports:

If I try to build it and run in in Docker and `-p 8000:8000 -p 3000:3000` to the `docker run` command. If I go to the url:3000/[page-route] I see the outline of the page, but I get an error, because the web socket traffic still can't get through.

Is there a way to get traffic to go to only one port. I have tried --

```
Connection Error

Cannot connect to server: websocket error. Check if server is reachable at http://localhost:8000
```

If I go to url:8000/[page-route], I get a page from the uvicorn reverse proxy with a caption [not found]. 

Is there a 'simple' way to deploy this that is stateless and all traffic is to and from the same port?

What am I missing here?

**Reply by masenf** [reply]
The API_URL arg is very important, as that must be the publicly accessible endpoint where the backend is listening. However, it can also be set as an environment variable at runtime, but only if you're running the actual nextjs server. For the docker build, the default behavior is to **export** the frontend to static files and only run the backend server (looks like the docs need to be updated a bit here).

If you removed the `RUN reflex export ...` line and remove `--backend-only` from the command, then you could make a stateless image and control the `API_URL` parameter at runtime.

---------

> Cannot connect to server: websocket error. Check if server is reachable at http://localhost:8000

This error arises because the default API_URL is `http://localhost:8000`. If you do not set API_URL, then the app cannot work in a real deployment because the client connecting to the frontend does not know the actual URL of the backend. We might be able to do something smarter here, like assuming the current URL and the backend are hosted on the same machine if only a port is provided, which would allow us to make the default API_URL something like `http://:8000`, which would probably work out of the box in more cases.

As for having all traffic go to the same port, our recommendation for that is to set up a reverse proxy, like `nginx` or `caddy` and have it forward the requests appropriately.

**Reply by david-thrower** [reply]
Here may be another isue: In k8s, for interrnal traffic, it will resolve to [service name].[namespace name].svc.cluster.local and externally as [whatever hostname the ingress is configured to]. Is there a way to just be able to omit this arg altogether?

Additionally, I don't have the `RUN reflex export ...`

... This is an example of what I have:

```
FROM python:3.11 as build
WORKDIR /app
COPY . . 
RUN pip install --upgrade pip

FROM build as runtime
WORKDIR /app
COPY requirements.txt .
RUN pip install -r requirements.txt

FROM runtime as final
WORKDIR /app/my-app
RUN /usr/local/bin/reflex init
# CMD reflex run --env dev --frontend-only


CMD ["reflex","run" , "--env", "dev"]
# Also tried everyu fathomable permutation of
# [CMD | ENTRYPOINT] ["reflex", "run", [--frontend-only]  [--env] [prod | dev]   ] 
# Have also configured with various permutations of frontend_port  backend_port  ...

```

**Reply by masenf** [reply]
Based on this discussion, I proposed https://github.com/reflex-dev/reflex/pull/1954, which will remove the need to set `api_url` explicitly, and will just use the same hostname that the frontend is running on (and the configured `backend_port`). If you're running a load balancer and combining the traffic to a single port, then you'll still need to specify API_URL to an address that the client browser can successfully connect to (usually with `https` and no port)

**Reply by david-thrower** [reply]
Another facet of the question: Is there a way to flat out disable the https functionality altogether. The service mesh and ingress will provide this anyway. Sorry to keep asking questions.

**Reply by masenf** [reply]
reflex does not handle https at all currently. if you need TLS, then you have to bring your own load balancer that terminates TLS

**Comment by david-thrower** [comment]
For the Edification of others: this worked:

Dockerfile:

```

FROM python:3.11 as build
WORKDIR /app/my-app
COPY . . 
RUN pip install --upgrade pip

FROM build as runtime
WORKDIR /app/my-app
COPY requirements.txt .
RUN pip install -r requirements.txt


FROM runtime as final
WORKDIR /app/my-app
RUN pwd
RUN ls


# CMD reflex run --env dev --frontend-only

ENTRYPOINT ["/bin/sh", "/app/my-app/entrypoint.sh"]

EXPOSE 3000
EXPOSE 8000
```

config-template.py:

```python

from os import getenv

API_URL = getenv("API_URL", default="")
if not API_URL:
    raise ValueError('The environment variable "API_URL" must be set. Try `docker run -e API_URL="http://foo-url:3000"`')

template = f"""
import reflex as rx


config = rx.Config(
    app_name="my-app",
    api_url="{API_URL}"
)

"""

with open('rxconfig.py', 'w') as f:
  f.write(template)

print("Wrote as rxconfig.py:")
print(template)
```

entrypoint.sh:

```shell
python3 /app/my-app/config-template.py
/usr/local/bin/reflex init
reflex run --env dev --backend-port 8000
```
To run:


```
docker run -it -p 3000:3000 -p 8000:8000 -e  API_URL="http://123.123.123.123:8000"" my-repo/my-container-image:my-tag
```
